File: leetcode-python-go/1138. 字母板上的路径/main.py
target = "code"

def alphabetBoardPath(target: str) -> str:
    ans = []
    x = y = 0
    for c in target:
        nx, ny = divmod(ord(c) - ord('a'), 5)  # 目标位置
        v = "UD"[nx > x] * abs(nx - x)  # 竖直
        h = "LR"[ny > y] * abs(ny - y)  # 水平
        ans.append((v + h if c != 'z' else h + v) + "!")
        x, y = nx, ny
    return ''.join(ans)

print(alphabetBoardPath(target))

File: leetcode-python-go/1138. 字母板上的路径/main.go
package main

import (
	"fmt"
	"strings"
)

func alphabetBoardPath(target string) string {
	ans := []byte{}
	x, y := 0, 0
	for _, c := range target {
		nx, ny := int(c-'a')/5, int(c-'a')%5                      // 目标位置
		v := strings.Repeat(string("UD"[b2i(nx > x)]), abs(nx-x)) // 竖直
		h := strings.Repeat(string("LR"[b2i(ny > y)]), abs(ny-y)) // 水平
		if c == 'z' {
			v, h = h, v
		}
		ans = append(ans, v+h+"!"...)
		x, y = nx, ny
	}
	return string(ans)
}

func b2i(b bool) int {
	if b {
		return 1
	}
	return 0
}
func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func main() {
	target := "code"
	results := alphabetBoardPath(target)
	fmt.Println(results)
}
